### ISSUE TITLE
New schema type: indexed string

### ISSUE THREAD

**Issue by larryli** [yiisoft/yii2#16489]
### What steps will reproduce the problem?

Use mysql 5.6 and utf8mb4.

```php
class mxxx_yyy_create_session_table extends Migration
{
    /**
     * @inheritdoc
     */
    public function safeUp()
    {
        $tableOptions = null;
        if ($this->db->driverName === 'mysql') {
            // http://stackoverflow.com/questions/766809/whats-the-difference-between-utf8-general-ci-and-utf8-unicode-ci
            // http://stackoverflow.com/questions/35125933/mysql-utf8mb4-errors-when-saving-emojis
            $tableOptions = 'CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci ENGINE=InnoDB';
        }

        $this->createTable('{{%session}}', [
            'id' => $this->string()->notNull(),
            'expire' => $this->integer()->notNull(),
            'data' => $this->binary()->notNull(),
            'PRIMARY KEY ([[id]])',
        ], $tableOptions);
    }

    /**
     * @inheritdoc
     */
    public function safeDown()
    {
        $this->dropTable('{{%session}}');
    }
}
```

### What is the expected result?

see https://stackoverflow.com/questions/1814532/1071-specified-key-was-too-long-max-key-length-is-767-bytes

```
    > create table {{%session}} ...Exception: SQLSTATE[42000]: Syntax error or access violation: 1071 Specified key was too long; max key length is 767 bytes
The SQL being executed was: CREATE TABLE `session` (
        `id` varchar(255) NOT NULL,
        `expire` int(11) NOT NULL,
        `data` blob NOT NULL,
        PRIMARY KEY (`id`)
) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci ENGINE=InnoDB (/vagrant/vendor/yiisoft/yii2/db/Schema.php:664)
```

### What do you get instead?

```php
 'id' => $this->string(180)->notNull(),
  // or
 'id' => 'varchar(255) CHARACTER SET latin1 COLLATE latin1_german1_ci NOT NULL',
```

So, add a new schema type, such as `indexedString()`:

```php
 'id' => $this->indexedString()->notNull(),
```

build:

```sql
`id` varchar(255) CHARACTER SET latin1 COLLATE latin1_german1_ci NOT NULL
```

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.x

**Comment by samdark** [comment]
Indeed, length of the index varies and default length for multibyte encoding doesn't work.

The feature itself is quite useless for regular projects but could be useful for installable products such as CMS.

**Comment by samdark** [comment]
Why 👎 , @rob006?

**Comment by rob006** [comment]
Supporting this at framework level does not make any sense. This issue is caused by mix of old and incorrectly configured DBMS and custom migration code. And the only sane and generic way to solve this is to use `string(190)`. So we need a wrapper for `string(190)` to support already fixed (this issue does not exist in MySQL 5.7+) bug in one DBMS?

**Comment by samdark** [comment]
Probably not.
